smart-contracts/contracts/cl-vault/src/rewards/distribution.rs

```diff
@@ -159,101 +81,188 @@ pub fn handle_collect_spread_rewards_reply(
+
+    let mut response_coin_list = CoinList::new();
+    response_coin_list.merge_osmocoins(data.collected_spread_rewards)?;
 
-    // calculate and update the strategist fee
+    // calculate the strategist fee and remove the share at source
     let vault_config = VAULT_CONFIG.load(deps.storage)?;
-    let strategist_fee = rewards.sub_ratio(vault_config.performance_fee)?;
+    let strategist_fee = response_coin_list.sub_ratio(vault_config.performance_fee)?;
     STRATEGIST_REWARDS.update(deps.storage, |old| old.add(strategist_fee))?;
 
+    let mut rewards = CURRENT_REWARDS.load(deps.storage)?;
+    rewards.update_rewards_coin_list(response_coin_list)?;
+
     CURRENT_REWARDS.save(deps.storage, &rewards)?;
 
-    Ok(Response::new().add_attribute(
-        "collected_spread_rewards",
-        format!("{:?}", response.collected_spread_rewards),
-    ))
+    // TODO add a nice response
+    Ok(Response::new()
+        .add_attribute("method", "reply")
+        .add_attribute("action", "handle_collect_spread_rewards"))
 }
 
-pub fn execute_distribute_rewards(
+pub fn execute_auto_compound_swap(
     deps: DepsMut,
-    _env: Env,
-    amount_of_users: Uint128,
+    env: Env,
+    force_swap_route: bool,
+    mut swap_routes: Vec<AutoCompoundAsset>,
 ) -> Result<Response, ContractError> {
-    let mut rewards_status = REWARDS_STATUS.load(deps.storage).unwrap();
-    if rewards_status != RewardsStatus::Distributing {
-        return Err(ContractError::IsNotDistributing {});
+    // auto compound admin
+    let dex_router = DEX_ROUTER.may_load(deps.storage)?;
+    if swap_routes.is_empty() {
+        return Err(ContractError::EmptyCompoundAssetList {});
     }
 
-    let rewards = CURRENT_REWARDS.load(deps.storage)?;
-    let total_shares = CURRENT_TOTAL_SUPPLY.load(deps.storage)?;
-    let mut distributed_rewards = DISTRIBUTED_REWARDS.load(deps.storage).unwrap();
-    let mut users_processed: u128 = 0;
-
-    while users_processed < amount_of_users.u128() {
-        // Attempt to pop a (user_address, share_amount) tuple from the front of the snapshot Deque
-        if let Some((user_address, share_amount)) = DISTRIBUTION_SNAPSHOT.pop_front(deps.storage)? {
-            // Calculate the user's reward based on the share amount directly
-            let reward_ratio = Decimal::from_ratio(share_amount, total_shares.u128());
-            let user_reward = rewards.mul_ratio(reward_ratio);
-            let user_reward_clone = user_reward.clone();
-
-            // Merge the new reward with any existing rewards for the user
-            USER_REWARDS.update(
-                deps.storage,
-                user_address,
-                |old| -> ContractResult<CoinList> {
-                    Ok(match old {
-                        Some(old_rewards) => user_reward_clone.add(old_rewards)?,
-                        None => user_reward_clone,
-                    })
+    let current_swap_route = swap_routes[0].clone();
```

Comment:
Probably here we want to enhance this handling. Looks like we miss a logic step which is detecting if any given route (either 1 or more) is related to any denom not included in the pool pair. So we want to swap those first i.e. N% N% to asset0 and asset1 or 100% asset0 or asset1 based on swap efficiency (we will have to rearrange percentages on the next re-range anyway so swap into a single asset is likely to be simpler).

Reply:
Instead of this we are pool them all together as sub messages array.